demos/yielding/README.md

```diff
@@ -0,0 +1,112 @@
+# Yielding
+
+In this demo, we try to show the usage of yielding, which is one characteristic of a multitasking system. 
+
+By calling syscall `yield()`, the thread gives away the utilization of a CPU to another thread, or a specific thread (will be discussed later in another demo).
+
+To achieve thread yielding, we need to introduce the following functions:
+
+* `add_thread_from_worker` (in `src/user/thread.c`)
+* `do_scheduler` (in `src/kernel/thread.c`)
+* `yield` (in `src/user/thread.c`)
+
+## `add_thread_from_worker`
+
+In AMT, we try to maintain all of the threads in an array called `all_threads`.
+
+By calling this function, the AMT will create a new thread instance with the given arguments, initialize the permissions, and put the function in. 
+
+After the creation of the thread, the entity will be put into `all_threads` array and wait for the utilization of the CPU.
+
+## `do_scheduler`
+
+The job of the scheduler is to choose the next thread to make use of the CPU. 
+
+In AMT, the scheduler simply walks through the `all_threads` array to find the next thread.
+
+To run the thread chosen by the scheduler and restore the function call stack of the current thread, we need to save the information of this thread first.
```

Comment:
Your description here isn't quite right. To be sure we're on the same page, here's how we get to do_scheduler most of the time: (first boot is a little different)
- user thread calls yield()
- this does syscall_yield
- we enter the kernel
- current thread context is saved
- we find that it's a syscall_yield and handle it
- by calling do_scheduler
- do_scheduler picks the next_thread
- we go back into assembly to switch the **already saved** current thread with the next one

There are times where yield.S will call do_scheduler but I don't think we need to cover them here. (I hope)